kolibri/plugins/facility/assets/src/views/users/sidePanels/AssignCoachesSidePanel.vue

```diff
@@ -1,35 +1,347 @@
+
+      const selectedUsersCount = computed(() => props.selectedUsers.size);
+
+      const hasSelectedClasses = computed(() => selectedClasses.value.length > 0);
+
+      const getSelectedClassObjects = computed(() =>
+        props.classes.filter(cls => selectedClasses.value.includes(cls.id)),
+      );
+
+      // Filter eligible users (coaches, admins, superusers)
+      const eligibleUsers = computed(() => {
+        return facilityUsers.value.filter(
+          user =>
+            props.selectedUsers.has(user.id) &&
+            (user.kind.includes(UserKinds.COACH) ||
+              user.kind === UserKinds.ADMIN ||
+              user.kind === UserKinds.SUPERUSER ||
+              user.is_superuser),
+        );
+      });
+
+      // Filter ineligible users (learners)
+      const ineligibleUsers = computed(() => {
+        return facilityUsers.value.filter(
+          user => props.selectedUsers.has(user.id) && user.kind === UserKinds.LEARNER,
+        );
+      });
+      const ineligibleUsersCount = computed(() => ineligibleUsers.value.length);
+
+      const eligibleUsersCount = computed(() => {
+        return eligibleUsers.value.length;
+      });
+
+      // Methods
+      async function handleAssign() {
+        if (!hasSelectedClasses.value) {
+          return;
+        }
+
+        isLoading.value = true;
+        showErrorWarning.value = false;
+        createdRoles.value = [];
+
+        try {
+          await assignCoachesToClasses();
+          createSnackbar(coachesAssignedNotice$());
+          showUndoModal.value = true;
+        } catch (error) {
+          showErrorWarning.value = true;
+        } finally {
+          isLoading.value = false;
+        }
+      }
+
+      async function assignCoachesToClasses() {
+        const selectedClasses = getSelectedClassObjects.value;
+        const eligibleUserIds = eligibleUsers.value.map(user => user.id);
+        if (eligibleUserIds.length === 0) {
+          return;
+        }
+
+        for (const classObj of selectedClasses) {
+          const newRoles = await RoleResource.saveCollection({
+            data: eligibleUserIds.map(userId => ({
+              collection: classObj.id,
+              user: userId,
+              kind: UserKinds.COACH,
+            })),
+          });
+          // Only add roles that were actually created (have an id)
+          const actuallyCreatedRoles = newRoles.filter(role => role.id);
+          createdRoles.value.push(...actuallyCreatedRoles);
+        }
+      }
+
+      function handleDismissConfirmation() {
+        showUndoModal.value = false;
+        instance.proxy.$emit('clearSelection');
+        instance.proxy.$router.back();
+      }
+
+      async function handleUndoAssignments() {
+        isLoading.value = true;
+        try {
+          if (createdRoles.value.length > 0) {
+            for (const role of createdRoles.value) {
+              if (role.id) {
+                await RoleResource.deleteModel({ id: role.id });
```

Comment:
(Mega important) Idem, we are making too many api request, we can bulk delete these roles using the `by_ids` filter. (Here its even more important since we are making a request for evey single combination of classes/users):

```javascript
RoleResource.deleteCollection({ by_ids })
```